File: jump_game_ii.cpp
#include "header.h"

int jump(int A[], int n) {
    if(n == 1) return 0;
    int num = 0;
    int l = 0,r=0;
    while(r<n-1){
        int max = r;
        for(int i=l;i<=r;i++)
            max = std::max(max,A[i]+i);
        l = r+1;
        r = max;
        num++;
    }
    return num;
}

int main(){
    int A[] = {2, 3, 1, 1, 4};
    int res = jump(A,sizeof(A)/sizeof(A[0]));
    printf("%d\n", res);
    return 0;
}

File: jump_game_ii.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        if len(A) == 0:
            return 0
        res = 0
        start = end = 0
        if len(A) == 0:
            return 0
        while end<len(A)-1:
            new_end = end
            for i in range(start,end+1,1):
                new_end = max(i+A[i],new_end)
            if new_end >= len(A)-1:
                return res+1
            res += 1
            start = end+1
            end = new_end


print Solution().jump([2, 3, 1, 1, 4])
